rocAL Serialize C++ Implementation

fiona-gladwin opened the pull request:
- Add support to serialize a simple pipeline, with Image loader and brightness
- Introduce protobuffers to store the information regarding the pipeline, ops, args, inputs and outputs
- Introduce PipelineOperator class in MasterGraph to store all the operators in the pipeline
- Introduce Argument class to store the details of each argument corresponding to each Node/Operator
- Add support to serialize all the pipeline information and serialize into a string

fiona-gladwin committed:
Add rocal proto file and CMake changes

fiona-gladwin committed:
Add initial API for rocal serialize

fiona-gladwin committed:
Add initial support for serialize test

fiona-gladwin committed:
Add initial changes to add operators and arguments in Master graph and Node

fiona-gladwin committed:
Add arguments in respective nodes for reader, image loader and brightness node

fiona-gladwin committed:
Add support to serialize args into protobuf

fiona-gladwin committed:
Update proto file

fiona-gladwin committed:
Introduce tensor name member to Tensor class

fiona-gladwin committed:
Add support to serialize Inputs and Outputs

fiona-gladwin committed:
Introduce member functions to obtain random param info

fiona-gladwin committed:
Add support for params in Arguments protobuf

fiona-gladwin committed:
Add support to serialize pipeline ouputs

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into fg/serialize

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into fg/serialize

fiona-gladwin committed:
Add comments for new functions

fiona-gladwin committed:
Introduce generic function to add all arguments

fiona-gladwin committed:
Use generic node argument setter for image loader

fiona-gladwin committed:
Remove commented code

fiona-gladwin committed:
Add changes for image loader single shard

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into fg/serialize

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into fg/serialize

fiona-gladwin committed:
Modify existing API to return size and allocate string based on the size

Introduce new API in master graph to return string
Add external API to copy the serialized string

fiona-gladwin committed:
Add python support for rocAL serialize

fiona-gladwin committed:
Add python example for pipeline serialize

fiona-gladwin committed:
Add support to serialize crop node and rotate nodes

fiona-gladwin committed:
Add support to store the serialized string in a file

fiona-gladwin committed:
Add changes in other init of crop and rotate nodes

fiona-gladwin committed:
Add support to process vectors in Argument and serialize it

fiona-gladwin committed:
Fix args list in ImageLoaderNode

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into fg/serialize

rrawther commented on rocAL/include/pipeline/node.h:
```diff
@@ -23,10 +23,190 @@ THE SOFTWARE.
+    std::string arg_name;   // Name of the argument
+    std::string type_name;  // Denotes the data type of the argument
+    std::string enum_type_name; // Denotes the name of the enum <arg_name_enum>
+    // TODO - Make an enum
```
Yes please make it as enum

rrawther commented on rocAL/include/pipeline/node.h:
```diff
@@ -23,10 +23,190 @@ THE SOFTWARE.
+    bool is_vector = false;
+    bool is_parameter = false;
+    bool is_null_ptr = false;
+    std::vector<std::any> values;   // Can change to std::variant later
```
what does this represent?

fiona-gladwin replied:
This denotes the value of the argument and it can take std::any type.
The argument can be a single value or a vector of values, hence we have used vector here

rrawther commented on rocAL/include/pipeline/node.h:
```diff
@@ -23,10 +23,190 @@ THE SOFTWARE.
+                enum_type_name = it->second;
+                values.push_back(static_cast<int>(val));
+            } else {
+                std::cout << "Type: Unknown" << arg_name << std::endl;
```
std::err is preferred for error outputs. Applies for all such

rrawther commented on rocAL/include/pipeline/node.h:
```diff
@@ -23,10 +23,190 @@ THE SOFTWARE.
+                type_name = it->second;                
+                is_vector = true;
+                for (const auto& v : val) {
+                    // values.push_back(v);  // Store std::string as std::any
```
remove comment

rrawther commented on rocAL/include/pipeline/node.h:
```diff
@@ -23,10 +23,190 @@ THE SOFTWARE.
+        is_vector = true;
+        if (!val.empty()) {
+            for (const auto& pair : val) {
+                values.push_back(static_cast<std::string>(pair.first));  // Push key
```
suggest storing key,value as std::pair as they always go together

fiona-gladwin replied:
This is a map<string, string> inorder to store that in the protobuffers, we are converting it to a single vector of string values, with keys first followed by values of the map

rrawther commented on rocAL/include/pipeline/node.h:
```diff
@@ -23,10 +23,190 @@ THE SOFTWARE.
+
+        // For MetadataReader case store an empty value
+        // During deserialization the MetadataReader should be created and passed from the MasterGraph.
+        if (name == "meta_data_reader") {
```
can't we pass the dafault value through constructor instead of hardcoding here

rrawther commented on rocAL/include/pipeline/node.h:
```diff
@@ -23,10 +23,190 @@ THE SOFTWARE.
+    explicit inline Argument(std::string name, std::string type, std::vector<T> &val)
+        : arg_name(name), type_name(type) {
+        is_vector = true;
+        for (const auto& v : val) {
```
why do we need a for loop. Can't you assign one vector to other or use copy constructor

fiona-gladwin replied:
We cannot assign a std::vector<T> to std::vector<std::any> directly as it would lead to compilation error

rrawther commented on rocAL/include/pipeline/node.h:
```diff
@@ -23,10 +23,190 @@ THE SOFTWARE.
+        : arg_name(name) {
+        type_name = "float";
+        if (param == nullptr) {
+            is_null_ptr = true;
```
why do you need this flag. type_name is sufficient for detecting? Also I suggest passing type_name as well in the constructor and not do anything inside the function. e.g: type_name_(type_name)

fiona-gladwin replied:
The user would only pass the argument and we internally deduce the type from the value passed, this way we can have a generic implementation to pass different args.

rrawther commented on rocAL_pybind/rocal_pybind.cpp:
```diff
@@ -294,6 +294,13 @@ PYBIND11_MODULE(rocal_pybind, m) {
     m.def("rocalRelease", &rocalRelease, py::return_value_policy::reference);
+    m.def("rocalSerialize", [](RocalContext context) {
+        size_t size;
+        rocalSerialize(context, size);
```
is there a return value from this function. What if it fails

rrawther commented on tests/cpp_api/serialize_tests/rocal_serialize_test.cpp:
```diff
@@ -0,0 +1,354 @@
+#include "rocal_api.h"
+using namespace cv;
+
+#if USE_OPENCV_4
```
check with Kiriti if we still need to support older versions of OpenCV

rrawther commented on tests/cpp_api/serialize_tests/rocal_serialize_test.cpp:
```diff
@@ -0,0 +1,354 @@
+
+    rocalSetSeed(0);
+
+    // // Creating uniformly distributed random objects to override some of the default augmentation parameters
```
remove commented code

rrawther commented on tests/cpp_api/serialize_tests/rocal_serialize_test.cpp:
```diff
@@ -0,0 +1,354 @@
+    std::cerr << serialized_pipe_string << "\n";
+    std::cerr << "==================================================================\n";
+    
+    /*>>>>>>>>>>>>>>>>>>> Diplay using OpenCV <<<<<<<<<<<<<<<<<*/
```
Assuming all commented code will be fixed before making PR into main branch

rrawther reviewed (changes requested):
Added review comments

fiona-gladwin committed:
Create new file for Pipeline operators

fiona-gladwin committed:
Remove common code

fiona-gladwin committed:
Minor fix - serialize

Add support to pass the reference to the size argument
Add is_vector and is_argument_input to the protobuf

fiona-gladwin committed:
Remove redundant code for args setting and format

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into fg/serialize

fiona-gladwin committed:
Introduce pipeline serializer class

fiona-gladwin committed:
Introduce EnumType in proto to handle enums

fiona-gladwin committed:
Use enums present in commons.h

Remove the usage of API enums in internal node files

fiona-gladwin committed:
Remove unused include rocal_api_types.h

fiona-gladwin committed:
Modify the type_names map to use internal enum only

fiona-gladwin committed:
Add enum registration method to register all internal enums

fiona-gladwin committed:
Introduce argument file in pipeline

fiona-gladwin committed:
Optimize Arguments file, remove unused constructors

fiona-gladwin committed:
Register metadata reader type and store acc in label reader

fiona-gladwin committed:
Remove RocalStatus return type

fiona-gladwin committed:
Modularise parameter serialization

fiona-gladwin committed:
Update rocal proto to use proto3 format

fiona-gladwin committed:
Introduce vectors in proto file

fiona-gladwin committed:
Merge branch 'fg/enum_change' into fg/serialize

fiona-gladwin committed:
Store param instead of param core

fiona-gladwin committed:
Modularize param retreival
